From a3abe5f8e8100ba6a0f982421b943135df34f27f Mon Sep 17 00:00:00 2001
From: Jitendra Pandey <jitendra@apache.org>
Date: Mon, 1 Jan 2018 21:51:48 -0800
Subject: [PATCH 2687/2700] HADOOP-15143. NPE due to Invalid KerberosTicket in
 UGI. Contributed by Mukul Kumar Singh.

Ref: CDH-64531

(cherry picked from commit cb60c05c3aff762c39d7d81df9a7b6e071acb181)

Change-Id: I6e477465b40a49b61e0cf183865bbf1a5165f40d
(cherry picked from commit e946cea7caa9d8cdf4d670b54ecc802fb11db7c6)
---
 .../hadoop/security/UserGroupInformation.java      |    5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/security/UserGroupInformation.java b/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/security/UserGroupInformation.java
index a5e6cd2..4ca47eb 100644
--- a/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/security/UserGroupInformation.java
+++ b/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/security/UserGroupInformation.java
@@ -1205,7 +1205,10 @@ private void fixKerberosTicketOrder() {
         Object cred = iter.next();
         if (cred instanceof KerberosTicket) {
           KerberosTicket ticket = (KerberosTicket) cred;
-          if (!ticket.getServer().getName().startsWith("krbtgt")) {
+          if (ticket.isDestroyed() || ticket.getServer() == null) {
+            LOG.warn("Ticket is already destroyed, remove it.");
+            iter.remove();
+          } else if (!ticket.getServer().getName().startsWith("krbtgt")) {
             LOG.warn(
                 "The first kerberos ticket is not TGT"
                     + "(the server principal is {}), remove and destroy it.",
-- 
1.7.9.5

